### PR TITLE
feat: make gen4 plan subsharding queries better

### PR DESCRIPTION
**Pull request by systay**
## Description
The gen4 planner was treating partial vindex lookups as equivalent to vindex lookups with all columns accounted for.
This lead to bad plans when some of the column comparisons were `IN` comparisons.

## Checklist
-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/testdata/dml_cases.txt [hunk 1]
```diff
@@ -3654,7 +3679,26 @@ Gen4 plan same as above
...
+      "Sharded": true
+    },
+    "TargetTabletType": "PRIMARY",
+    "MultiShardAutocommit": false,
+    "Query": "update multicol_tbl set x = 1 where `name` = 'foo' and cola = 2",
+    "Table": "multicol_tbl",
+    "Values": [
+      "VARCHAR(\"foo\")"
+    ],
+    "Vindex": "name_muticoltbl_map"
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/dml_cases.txt:3699]: Not sure that a lookup vindex is a better plan than doing a SubShard query against the primary vindex, but this should be addressed separately, imho.

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/engine/routing.go [hunk 2]
```diff
@@ -68,8 +68,8 @@ const (
 	// Is used when the query explicitly sets a target destination:
 	// in the clause e.g: UPDATE `keyspace[-]`.x1 SET foo=1
 	ByDestination
-	// NumOpcodes is the number of opcodes
-	NumOpcodes
+	// SubShard is for when we are missing one or more columns from a composite vindex
```
Comment (harshit-gangal) [inline go/vt/vtgate/engine/routing.go:71]: adding it here does not mean that Scatter is better than SubShard?

File: go/vt/vtgate/planbuilder/route_test.go [hunk 3]
```diff
@@ -28,22 +28,26 @@ import (
...
+	Equal       	 2
+	IN          	 3
+	MultiEqual  	 4
+	Scatter     	 5
+	Next        	 6
+	DBA         	 7
+	Reference   	 8
+	None        	 9
+	Subshard         10 <- not covered
+	NumRouteOpcodes  11
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/route_test.go:46]: NumRouteOpcodes is removed from opcode list

**Review by harshit-gangal — COMMENTED** [review]
left some comments

**Pull request merged**